exception when changing the number of dataset using BalancedPathFilter

I want to control the dataset size using BalancedPathFilter(new Random(seed), labelMaker, 20000, numLabels, 6000).The total dataset size is 60000,but I only want to use 20000 to train.
However,
java.lang.RuntimeException: java.lang.UnsupportedOperationException: Cannot do conversion to one hot using batched reader: 10 output classes, but array.size(1) is 4 (must be equal to 1 or numClasses = 10)
        at org.deeplearning4j.datasets.iterator.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:430)
     Caused by: java.lang.UnsupportedOperationException: Cannot do conversion to one hot using batched reader: 10 output classes, but array.size(1) is 4 (must be equal to 1 or numClasses = 10)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.convertWritablesBatched(RecordReaderMultiDataSetIterator.java:413)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.convertFeaturesOrLabels(RecordReaderMultiDataSetIterator.java:359)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.nextMultiDataSet(RecordReaderMultiDataSetIterator.java:332)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.next(RecordReaderMultiDataSetIterator.java:212)
        at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:364)
        at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:439)
        at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:84)
        at org.deeplearning4j.datasets.iterator.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:404)

How can I control the dataset size for training

just the MnistClassifier example
```
 val trainData = File(rootDir + "mnist_png/training")
 val randNumGen = Random(1234)
 val trainSplit = FileSplit(trainData, NativeImageLoader.ALLOWED_FORMATS, Random(1234))
 val labelMaker = ParentPathLabelGenerator() // use parent directory name as the image label
 val pathFilter = BalancedPathFilter(Random(42),labelMaker,20000,10,6000)
 val split = trainSplit.sample(pathFilter,1.0)
 val trainRR = ImageRecordReader(28, 28, 1, labelMaker)

 trainRR.initialize(split[0])
 samplesUsedInTraining = split[0].length().toInt()

 val trainIter = RecordReaderDataSetIterator(trainRR, 54, 1, 10)

 // pixel values from 0-255 to 0-1 (min-max scaling)
 val imageScaler = ImagePreProcessingScaler()
 imageScaler.fit(trainIter)
 trainIter.preProcessor = imageScaler
 model!!.fit(trainIter)
```
the errors are
```
java.lang.RuntimeException: java.lang.UnsupportedOperationException: Cannot do conversion to one hot using batched reader: 10 output classes, but array.size(1) is 4 (must be equal to 1 or numClasses = 10)
        at org.deeplearning4j.datasets.iterator.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:430)
     Caused by: java.lang.UnsupportedOperationException: Cannot do conversion to one hot using batched reader: 10 output classes, but array.size(1) is 4 (must be equal to 1 or numClasses = 10)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.convertWritablesBatched(RecordReaderMultiDataSetIterator.java:413)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.convertFeaturesOrLabels(RecordReaderMultiDataSetIterator.java:359)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.nextMultiDataSet(RecordReaderMultiDataSetIterator.java:332)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.next(RecordReaderMultiDataSetIterator.java:212)
        at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:364)
        at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:439)
        at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:84)
        at org.deeplearning4j.datasets.iterator.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:404)
```
when I print the trainRR.numLabels(),it is less than 10.But how can I solve it?

RandomPathFilter has the same issue
```
 val trainData = File(rootDir + "mnist_png/training")
 val randNumGen = Random(1234)
 val trainSplit = FileSplit(trainData, NativeImageLoader.ALLOWED_FORMATS, Random(1234))
 val labelMaker = ParentPathLabelGenerator() // use parent directory name as the image label
 val pathFilter = RandomPathFilter(Random(42),labelMaker,20000)
 val split = trainSplit.sample(pathFilter,1.0)
 val trainRR = ImageRecordReader(28, 28, 1, labelMaker)

 trainRR.initialize(split[0])
 samplesUsedInTraining = split[0].length().toInt()

 val trainIter = RecordReaderDataSetIterator(trainRR, 54, 1, 10)

 // pixel values from 0-255 to 0-1 (min-max scaling)
 val imageScaler = ImagePreProcessingScaler()
 imageScaler.fit(trainIter)
 trainIter.preProcessor = imageScaler
 model!!.fit(trainIter)
```
```
java.lang.RuntimeException: java.lang.UnsupportedOperationException: Cannot do conversion to one hot using batched reader: 10 output classes, but array.size(1) is 4 (must be equal to 1 or numClasses = 10)
        at org.deeplearning4j.datasets.iterator.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:430)
     Caused by: java.lang.UnsupportedOperationException: Cannot do conversion to one hot using batched reader: 10 output classes, but array.size(1) is 4 (must be equal to 1 or numClasses = 10)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.convertWritablesBatched(RecordReaderMultiDataSetIterator.java:413)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.convertFeaturesOrLabels(RecordReaderMultiDataSetIterator.java:359)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.nextMultiDataSet(RecordReaderMultiDataSetIterator.java:332)
        at org.deeplearning4j.datasets.datavec.RecordReaderMultiDataSetIterator.next(RecordReaderMultiDataSetIterator.java:212)
        at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:364)
        at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:439)
        at org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator.next(RecordReaderDataSetIterator.java:84)
        at org.deeplearning4j.datasets.iterator.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:404)
```

--- a/datavec/datavec-api/src/main/java/org/datavec/api/io/filters/RandomPathFilter.java
+++ b/datavec/datavec-api/src/main/java/org/datavec/api/io/filters/RandomPathFilter.java
@@ -17,6 +17,7 @@
 package org.datavec.api.io.filters;
 
 import java.net.URI;
+import java.util.Arrays;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Random;
@@ -64,8 +65,12 @@ public class RandomPathFilter implements PathFilter {
 
     @Override
     public URI[] filter(URI[] paths) {
+        // shuffle before to avoid sampling bias
+        ArrayList<URI> paths2 = new ArrayList<URI>(Arrays.asList(paths));
+        Collections.shuffle(paths2, random);
+
         ArrayList<URI> newpaths = new ArrayList<URI>();
-        for (URI path : paths) {
+        for (URI path : paths2) {
             if (accept(path.toString())) {
                 newpaths.add(path);
             }
@@ -73,7 +78,6 @@ public class RandomPathFilter implements PathFilter {
                 break;
             }
         }
-        Collections.shuffle(newpaths, random);
         return newpaths.toArray(new URI[newpaths.size()]);
     }
 }

--- a/datavec/datavec-api/src/test/java/org/datavec/api/split/InputSplitTests.java
+++ b/datavec/datavec-api/src/test/java/org/datavec/api/split/InputSplitTests.java
@@ -24,6 +24,7 @@ import org.datavec.api.io.labels.PatternPathLabelGenerator;
 import org.junit.Test;
 
 import java.io.*;
+import java.net.URI;
 import java.net.URISyntaxException;
 import java.util.ArrayList;
 import java.util.Random;
@@ -127,6 +128,27 @@ public class InputSplitTests {
         assertTrue(boostrap.needsBootstrapForWrite());
         boostrap.bootStrapForWrite();
         assertTrue(tmpDir.listFiles() != null);
+    }
+
+    @Test
+    public void testSampleNoBias() throws URISyntaxException {
+        Random random = new Random(7);
+        RandomPathFilter randomPathFilter = new RandomPathFilter(random, null, 100);
+
+        URI[] paths = new URI[1000];
+        for (int i = 0; i < paths.length; i++) {
+            paths[i] = new URI("file:///label" + (i/100) + "/image" + i);
+        }
+
+        boolean notOnlyFirstLabel = false;
+        URI[] paths2 = randomPathFilter.filter(paths);
+        assertEquals(100, paths2.length);
+        for (int i = 0; i < paths2.length; i++) {
+            if (!paths2[i].toString().startsWith("file:///label0/")) {
+                notOnlyFirstLabel = true;
+            }
+        }
+        assertTrue(notOnlyFirstLabel);
     }
 
 }

